src/flask_pyoidc/flask_pyoidc.py

```diff
@@ -319,3 +333,203 @@ def valid_access_token(self, force_refresh=False):
+        ------
+        NotForMe
+            If access_token is invalid.
+
+        """
+        received_access_token = self._parse_access_token(request)
+        # send token introspection request
+        result = client._token_introspection_request(
+            access_token=received_access_token)
+        logger.debug(result)
+        # Check if access_token is valid, active can be True or False
+        if not result.get('active'):
+            return False
+        # Check if client_id is in audience claim
+        if client._client.client_id not in result['aud']:
+            # log the exception if client_id is not in audience and returns
+            # False, you can configure audience with Identity Provider
+            logger.info('Token is valid but required audience is missing')
+            return False
+        # Check if the scopes associated with the access_token are the ones
+        # required by the endpoint and not something else which is not
+        # permitted.
+        if scopes and not set(scopes).issubset(set(result['scope'])):
+            logger.info('Token is valid but does not have required scopes')
+            return False
+        return result
+
+    def token_auth(self, provider_name, scopes_required: list = None):
+        """Token based authorization.
+
+        Parameters
+        ----------
+        provider_name : str
+            Name of the provider registered with OIDCAuthorization.
+        scopes_required : list, optional
+            List of valid scopes associated with the endpoint.
+
+        Raises
+        ------
+        flask.abort(401)
+            If authorization field is missing.
+        flask.abort(403)
+            If the access token is invalid.
+
+        How To Use
+        ----------
+        >>> auth = OIDCAuthentication({'default': provider_config})
+        >>> @app.route('/')
+            @auth.token_auth(provider_name='default')
+            def index():
+                ...
+        >>> # You can also specify scopes required by your endpoint.
+        >>> @auth.token_auth(provider_name='default',
+                             scopes_required=['read', 'write'])
+        """
+        def token_decorator(view_func):
+
+            @functools.wraps(view_func)
+            def wrapper(*args, **kwargs):
+
+                client = self.clients[provider_name]
+                # Check for authorization field in the request header.
+                if not self._check_authorization_header(flask.request):
+                    logger.info('Request header has no authorization field')
+                    # Abort the request if authorization field is missing.
+                    flask.abort(401)
+                token_introspection_result = self.introspect_token(
+                    request=flask.request, client=client,
+                    scopes=scopes_required)
+                if token_introspection_result:
+                    logger.info('user has valid access token')
```

Comment:
Nit: The token might be issued to a service instead of user, so more appropriate message would be something like "Request has valid access token".